container-search/src/main/java/com/yahoo/search/dispatch/rpc/RpcPing.java

```diff
@@ -33,32 +31,41 @@ public RpcPing(Node node, ClusterMonitor<Node> clusterMonitor, RpcResourcePool r
     }
 
     @Override
-    public Pong call() throws Exception {
+    public void ping(PongHandler handler) {
         try {
-            var queue = new LinkedBlockingQueue<ResponseOrError<ProtobufResponse>>(1);
-
-            sendPing(queue);
-
-            var responseOrError = queue.poll(clusterMonitor.getConfiguration().getRequestTimeout(), TimeUnit.MILLISECONDS);
-            if (responseOrError == null) {
-                return new Pong(ErrorMessage.createNoAnswerWhenPingingNode("Timed out waiting for pong from " + node));
-            } else if (responseOrError.error().isPresent()) {
-                return new Pong(ErrorMessage.createBackendCommunicationError(responseOrError.error().get()));
+            if (node.sendPing()) {
+                sendPing(handler);
             }
+        } catch (RuntimeException e) {
+            handler.handle(new Pong(
+                    ErrorMessage.createBackendCommunicationError("Exception when pinging " + node + ": " + Exceptions.toMessageString(e))));
+            node.receivePing();
+        } catch (Throwable throwable) {
```

Comment:
better to use finally than to hide Errors

Reply:
Take another look, Rewritten to use sequence numbers instead. Simpler and safer.